### ISSUE TITLE
Rule Change: support "object spread" boolean context

### ISSUE THREAD

**Issue by abrahamguo** [eslint/eslint#21328]
### What rule do you want to change?

no-extra-boolean-cast

### What change do you want to make?

Generate more warnings

### How do you think the change should be implemented?

Other

### Example code

```js
Boolean(a ?? !!x)            // ✅ flagged  — "Redundant double negation"
const o = { ...(!!x && a) }  // ❌ not flagged — but the !! is just as redundant
```

### What does the rule currently do for this code?

When `enforceForInnerExpressions` is `true`, it checks the RHS of `??` and correctly determines that the `!!` is unnecessary, and reports on line 1, but it does NOT check the LHS of `&&`.

### What will the rule do after it's changed?

When `enforceForInnerExpressions` is `true`, it will also check the LHS of a `&&` inside an object spread, as a "boolean context".

### Participation

- [x] I am willing to submit a pull request to implement this change.

### AI acknowledgment

- [x] I did not use AI to generate this issue report.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

**Comment by DMartens** [comment]
I agree that we should handle the left operand of `&&` inside an array or object spread as a boolean context.
But this would be a breaking change as it flags more code and is not a bugfix,
so this change would need to wait for the next major version.
I would like more opinions on this from the team on whether we should accept this proposal.

**Comment by mdjermanovic** [comment]
The `no-extra-boolean-cast` rule targets contexts where the value is coerced to a boolean so there's no need to convert it explicitly. In the case of object spread, values are not coerced to a boolean. Therefore, I don't think it qualifies for inclusion in this rule. The `!!` in `{ ...(!!x && a) }` indeed doesn't affect the result, but due to different reasons: the spread argument can evaluate to `x` only if `x` is falsy, and all falsy values are effectively no-op in that context. I think this is a specific case and logic that is beyond the scope of this rule.

**Comment by mdjermanovic** [comment]
Oh, and I forgot that this rule is marked as frozen (https://eslint.org/docs/latest/rules/no-extra-boolean-cast), so we are not accepting new feature requests for it.
